File: optimization-mathematical-algorithms-problems-optimized/src/problems.cpp
```cpp
#include "problems.h"
#include "utilities.h" // For MathUtils, potentially others

#include <numeric> // For std::gcd (C++17) - though we implement our own
#include <cmath>   // For std::abs, std::pow

// --- Problem 1: Greatest Common Divisor (GCD) & Least Common Multiple (LCM) ---

namespace GCD_LCM {

    /**
     * @brief Calculates the Greatest Common Divisor (GCD) of two numbers using a naive approach.
     *        It iterates from min(a,b) down to 1, checking for divisibility.
     * @param a The first non-negative integer.
     * @param b The second non-negative integer.
     * @return The GCD of a and b. Returns 0 if both a and b are 0 (undefined or convention).
     *         Handles negative inputs by taking their absolute values.
     * @complexity
     *   Time: O(min(a,b)) in the worst case (e.g., if a and b are coprime).
     *   Space: O(1)
     */
    long long gcd_naive(long long a, long long b) {
        // Handle negative inputs: GCD is typically defined for non-negative integers.
        // We take the absolute values as GCD(a, b) = GCD(|a|, |b|).
        a = MathUtils::absolute(a);
        b = MathUtils::absolute(b);

        if (a == 0 && b == 0) return 0; // Convention: GCD(0,0) is often undefined or 0.
        if (a == 0) return b;
        if (b == 0) return a;

        long long result = 1;
        for (long long i = std::min(a, b); i >= 1; --i) {
            if (a % i == 0 && b % i == 0) {
                result = i;
                break;
            }
        }
        return result;
    }

    /**
     * @brief Calculates the Greatest Common Divisor (GCD) of two numbers using the Euclidean Algorithm (recursive).
     *        Based on the property: GCD(a, b) = GCD(b, a % b). Base case: GCD(a, 0) = a.
     * @param a The first non-negative integer.
     * @param b The second non-negative integer.
     * @return The GCD of a and b.
     *         Handles negative inputs by taking their absolute values.
     * @complexity
     *   Time: O(log(min(a,b))) - The number of recursive calls is logarithmic with respect to the smaller number.
     *   Space: O(log(min(a,b))) - Due to the recursion stack depth.
     */
    long long gcd_euclidean_recursive(long long a, long long b) {
        // Handle negative inputs: GCD is typically defined for non-negative integers.
        // We take the absolute values as GCD(a, b) = GCD(|a|, |b|).
        a = MathUtils::absolute(a);
        b = MathUtils::absolute(b);

        // Base case: if b is 0, then a is the GCD
        if (b == 0) {
            return a;
        }
        // Recursive step: GCD(a, b) = GCD(b, a % b)
        return gcd_euclidean_recursive(b, a % b);
    }

    /**
     * @brief Calculates the Greatest Common Divisor (GCD) of two numbers using the Euclidean Algorithm (iterative).
     *        Uses the property: GCD(a, b) = GCD(b, a % b).
     * @param a The first non-negative integer.
     * @param b The second non-negative integer.
     * @return The GCD of a and b.
     *         Handles negative inputs by taking their absolute values.
     * @complexity
     *   Time: O(log(min(a,b))) - The number of iterations is logarithmic with respect to the smaller number.
     *   Space: O(1) - Constant extra space.
     */
    long long gcd_euclidean_iterative(long long a, long long b) {
        // Handle negative inputs: GCD is typically defined for non-negative integers.
        // We take the absolute values as GCD(a, b) = GCD(|a|, |b|).
        a = MathUtils::absolute(a);
        b = MathUtils::absolute(b);

        // Iterate until b becomes 0
        while (b != 0) {
            long long temp = b;
            b = a % b;
            a = temp;
        }
        return a; // When b is 0, a holds the GCD
    }

    /**
     * @brief Calculates the Least Common Multiple (LCM) of two numbers.
     *        Uses the property: LCM(a, b) = (|a * b|) / GCD(a, b).
     * @param a The first integer.
     * @param b The second integer.
     * @return The LCM of a and b. Returns 0 if either a or b is 0.
     *         Handles negative inputs by taking their absolute values.
     * @complexity
     *   Time: O(log(min(a,b))) - Dominated by the GCD calculation.
     *   Space: O(1) - Constant extra space (or O(log) if recursive GCD is used).
     */
    long long lcm(long long a, long long b) {
        // LCM is typically defined for positive integers. Taking absolute values for consistency.
        a = MathUtils::absolute(a);
        b = MathUtils::absolute(b);

        if (a == 0 || b == 0) {
            return 0; // LCM of any number with 0 is 0.
        }

        // Using the formula: LCM(a, b) = (a * b) / GCD(a, b)
        // To prevent overflow, we can compute (a / GCD(a, b)) * b.
        return (a / gcd_euclidean_iterative(a, b)) * b;
    }

} // namespace GCD_LCM


// --- Problem 2: Prime Number Generation (Sieve of Eratosthenes) ---

namespace PrimeNumbers {

    /**
     * @brief Checks if a number is prime using a naive trial division method.
     *        It iterates from 2 up to the square root of n.
     * @param n The integer to check for primality.
     * @return True if n is prime, false otherwise.
     * @complexity
     *   Time: O(sqrt(n))
     *   Space: O(1)
     */
    bool is_prime_naive(int n) {
        if (n <= 1) return false; // Numbers less than or equal to 1 are not prime
        if (n <= 3) return true;  // 2 and 3 are prime
        if (n % 2 == 0 || n % 3 == 0) return false; // Multiples of 2 or 3 are not prime

        // Check for factors from 5 onwards
        // All primes greater than 3 can be expressed in the form 6k ± 1
        for (int i = 5; i * i <= n; i = i + 6) {
            if (n % i == 0 || n % (i + 2) == 0) {
                return false;
            }
        }
        return true;
    }

    /**
     * @brief Generates all prime numbers up to a given limit using the Sieve of Eratosthenes.
     *        Marks multiples of each prime as not prime.
     * @param limit The upper bound (inclusive) for finding primes.
     * @return A boolean vector where `result[i]` is true if `i` is prime, false otherwise.
     * @complexity
     *   Time: O(N log log N) - Very efficient for finding all primes up to N.
     *   Space: O(N) - To store the boolean array.
     */
    std::vector<bool> sieve_of_eratosthenes(int limit) {
        if (limit < 0) return {}; // Handle invalid limit
        std::vector<bool> is_prime(limit + 1, true); // Initialize all numbers as potentially prime

        if (limit >= 0) is_prime[0] = false; // 0 is not prime
        if (limit >= 1) is_prime[1] = false; // 1 is not prime

        // Start from 2. If i is prime, mark all its multiples as not prime.
        // We only need to check up to sqrt(limit) because any composite number n
        // must have at least one prime factor less than or equal to sqrt(n).
        for (int p = 2; p * p <= limit; ++p) {
            // If is_prime[p] is still true, then it is a prime
            if (is_prime[p]) {
                // Mark all multiples of p (starting from p*p) as not prime.
                // Multiples less than p*p would have already been marked by smaller primes.
                for (int i = p * p; i <= limit; i += p) {
                    is_prime[i] = false;
                }
            }
        }
        return is_prime;
    }

    /**
     * @brief Generates a list of prime numbers up to a given limit using the Sieve of Eratosthenes.
     *        This function is a wrapper around `sieve_of_eratosthenes` to return a list of primes.
     * @param limit The upper bound (inclusive) for finding primes.
     * @return A vector containing all prime numbers up to 'limit'.
     * @complexity
     *   Time: O(N log log N) - Dominated by the sieve generation.
     *   Space: O(N) - For the boolean sieve array, plus O(pi(N)) for the result list (where pi(N) is the prime-counting function, approximately N/log N).
     */
    std::vector<int> get_primes_list_sieve(int limit) {
        std::vector<bool> is_prime = sieve_of_eratosthenes(limit);
        std::vector<int> primes_list;
        for (int p = 2; p <= limit; ++p) {
            if (is_prime[p]) {
                primes_list.push_back(p);
            }
        }
        return primes_list;
    }

} // namespace PrimeNumbers

// --- Problem 3: Power (x^n) Calculation ---

namespace PowerCalculation {

    /**
     * @brief Calculates x raised to the power of n using a naive iterative approach.
     *        Multiplies base 'n' times. Handles negative 'n' by taking reciprocal.
     * @param base The base number.
     * @param exponent The integer exponent.
     * @return The result of base^exponent.
     * @complexity
     *   Time: O(|n|) - Linear with respect to the absolute value of the exponent.
     *   Space: O(1)
     */
    double power_naive(double base, int exponent) {
        if (base == 0.0) {
            if (exponent == 0) return 1.0; // Convention: 0^0 is 1
            return 0.0; // 0^positive = 0, 0^negative is undefined (return 0.0 for practical purposes or error)
        }
        if (exponent == 0) return 1.0;

        double result = 1.0;
        long long abs_exponent = MathUtils::absolute(static_cast<long long>(exponent)); // Use long long for absolute value to handle INT_MIN

        for (long long i = 0; i < abs_exponent; ++i) {
            result *= base;
        }

        return (exponent < 0) ? (1.0 / result) : result;
    }

    /**
     * @brief Calculates x raised to the power of n using Binary Exponentiation (recursive).
     *        Also known as Exponentiation by Squaring. Efficiently computes x^n by halving the exponent.
     * @param base The base number.
     * @param exponent The integer exponent.
     * @return The result of base^exponent.
     * @complexity
     *   Time: O(log |n|) - Each recursive call roughly halves the exponent.
     *   Space: O(log |n|) - Due to the recursion stack depth.
     */
    double power_binary_exponentiation_recursive(double base, int exponent) {
        // Handle edge cases
        if (base == 0.0) {
            if (exponent == 0) return 1.0; // Convention: 0^0 is 1
            return 0.0; // 0^positive = 0, 0^negative is undefined (return 0.0 for practical purposes or error)
        }
        if (exponent == 0) return 1.0;
        if (base == 1.0) return 1.0; // 1 to any power is 1

        double result = 1.0;
        long long exp_abs = exponent; // Use long long to handle INT_MIN safely
        bool is_negative_exp = false;

        if (exp_abs < 0) {
            is_negative_exp = true;
            exp_abs = MathUtils::absolute(exp_abs); // Convert to positive exponent
        }

        // Recursive helper function
        // Note: For large exponents, passing `exp_abs` as `long long` is safer.
        // But for `int` exponent as per function signature, it fits within `int`.
        // The cast to `long long` for `exp_abs` is primarily for `INT_MIN` case
        // where `abs(INT_MIN)` is `INT_MIN` in 2's complement systems.
        // `MathUtils::absolute` correctly handles this.
        std::function<double(double, long long)> calculate_power =
            [&](double b, long long e) -> double {
            if (e == 0) return 1.0;
            double half_power = calculate_power(b, e / 2);
            if (e % 2 == 0) {
                return half_power * half_power;
            } else {
                return half_power * half_power * b;
            }
        };

        result = calculate_power(base, exp_abs);

        return is_negative_exp ? (1.0 / result) : result;
    }

    /**
     * @brief Calculates x raised to the power of n using Binary Exponentiation (iterative).
     *        Efficiently computes x^n by repeatedly squaring the base and adjusting the result.
     * @param base The base number.
     * @param exponent The integer exponent.
     * @return The result of base^exponent.
     * @complexity
     *   Time: O(log |n|) - Iterates through the bits of the exponent.
     *   Space: O(1) - Constant extra space.
     */
    double power_binary_exponentiation_iterative(double base, int exponent) {
        // Handle edge cases
        if (base == 0.0) {
            if (exponent == 0) return 1.0; // Convention: 0^0 is 1
            return 0.0; // 0^positive = 0, 0^negative is undefined (return 0.0 for practical purposes or error)
        }
        if (exponent == 0) return 1.0;
        if (base == 1.0) return 1.0; // 1 to any power is 1

        double result = 1.0;
        long long exp_abs = exponent; // Use long long to handle INT_MIN safely
        bool is_negative_exp = false;

        if (exp_abs < 0) {
            is_negative_exp = true;
            exp_abs = MathUtils::absolute(exp_abs); // Convert to positive exponent
        }

        double current_base = base;
        while (exp_abs > 0) {
            // If the current bit of exponent is 1 (i.e., exponent is odd)
            if (exp_abs % 2 == 1) {
                result *= current_base;
            }
            // Square the base for the next iteration (next bit)
            current_base *= current_base;
            // Halve the exponent (shift right by 1 bit)
            exp_abs /= 2;
        }

        return is_negative_exp ? (1.0 / result) : result;
    }

} // namespace PowerCalculation

// --- Problem 4: Nth Fibonacci Number ---

namespace Fibonacci {

    /**
     * @brief Calculates the Nth Fibonacci number using a naive recursive approach.
     *        F(n) = F(n-1) + F(n-2) with F(0)=0, F(1)=1.
     * @param n The index of the Fibonacci number to calculate (n >= 0).
     * @return The Nth Fibonacci number. Returns -1 for invalid input (n < 0).
     * @complexity
     *   Time: O(2^n) - Exponential due to redundant calculations (e.g., F(2) is calculated multiple times for F(5)).
     *   Space: O(n) - Due to the maximum depth of the recursion stack.
     */
    long long fibonacci_recursive_naive(int n) {
        if (n < 0) return -1; // Invalid input
        if (n == 0) return 0;
        if (n == 1) return 1;

        return fibonacci_recursive_naive(n - 1) + fibonacci_recursive_naive(n - 2);
    }

    /**
     * @brief Helper function for recursive Fibonacci with Memoization.
     *        Hides the memoization table from the public API.
     * @param n The index of the Fibonacci number to calculate.
     * @param memo_table A reference to the memoization vector.
     * @return The Nth Fibonacci number.
     */
    long long fibonacci_recursive_memoized_helper(int n, std::vector<long long>& memo_table) {
        if (n < 0) return -1; // Invalid input
        if (n == 0) return 0;
        if (n == 1) return 1;

        // If the result is already computed, return it
        if (memo_table[n] != -1) { // -1 indicates not computed
            return memo_table[n];
        }

        // Compute and store the result
        memo_table[n] = fibonacci_recursive_memoized_helper(n - 1, memo_table) +
                        fibonacci_recursive_memoized_helper(n - 2, memo_table);
        return memo_table[n];
    }

    /**
     * @brief Calculates the Nth Fibonacci number using recursive approach with Memoization (Dynamic Programming).
     *        Avoids redundant calculations by storing previously computed results.
     * @param n The index of the Fibonacci number to calculate (n >= 0).
     * @return The Nth Fibonacci number. Returns -1 for invalid input (n < 0).
     * @complexity
     *   Time: O(n) - Each Fibonacci number from 2 to n is computed only once.
     *   Space: O(n) - For the memoization table and O(n) for the recursion stack.
     */
    long long fibonacci_recursive_memoized(int n) {
        if (n < 0) return -1;
        // Initialize memoization table with -1 (or some indicator for not computed)
        // Size n+1 to store F(0) through F(n)
        std::vector<long long> memo_table(n + 1, -1);
        return fibonacci_recursive_memoized_helper(n, memo_table);
    }

    /**
     * @brief Calculates the Nth Fibonacci number using an iterative approach (Bottom-Up Dynamic Programming).
     *        Builds up the sequence from F(0) and F(1).
     * @param n The index of the Fibonacci number to calculate (n >= 0).
     * @return The Nth Fibonacci number. Returns -1 for invalid input (n < 0).
     * @complexity
     *   Time: O(n) - Linear, as it iterates from 2 up to n.
     *   Space: O(1) - Only a few variables are used to store the last two Fibonacci numbers.
     */
    long long fibonacci_iterative(int n) {
        if (n < 0) return -1; // Invalid input
        if (n == 0) return 0;
        if (n == 1) return 1;

        long long a = 0; // Represents F(i-2)
        long long b = 1; // Represents F(i-1)
        long long current_fib = 0; // Represents F(i)

        // Iterate from 2 up to n
        for (int i = 2; i <= n; ++i) {
            current_fib = a + b;
            a = b; // Update a to be the previous b
            b = current_fib; // Update b to be the current F(i)
        }
        return current_fib;
    }

} // namespace Fibonacci
```

File: coding-interview-mathematical-algorithms-plus/src/problems.js
```javascript
/**
 * src/problems.js
 *
 * This file contains optimal and efficient implementations for various mathematical problems.
 * Each function includes detailed comments, time/space complexity analysis, and explanations.
 */

// Import any necessary helpers
const { factorial } = require('./utils');

/**
 * Problem 1: Greatest Common Divisor (GCD) using Euclidean Algorithm
 *
 * The GCD of two integers is the largest positive integer that divides both numbers
 * without leaving a remainder.
 *
 * Optimal Approach: Euclidean Algorithm (Iterative)
 * The algorithm is based on the principle that the greatest common divisor of two
 * numbers does not change if the larger number is replaced by its difference with
 * the smaller number. This process continues until one of the numbers is zero,
 * and the other number is the GCD. More efficiently, it uses the remainder:
 * gcd(a, b) = gcd(b, a % b)
 *
 * @param {number} a The first non-negative integer.
 * @param {number} b The second non-negative integer.
 * @returns {number} The greatest common divisor of a and b.
 *
 * Time Complexity: O(log(min(a, b))) - The number of steps is logarithmic with respect to the smaller of the two numbers.
 *                   This is because the remainder decreases significantly in each step.
 * Space Complexity: O(1) - Uses a constant amount of extra space.
 */
function gcd(a, b) {
    // Ensure inputs are non-negative integers
    if (typeof a !== 'number' || typeof b !== 'number' || !Number.isInteger(a) || !Number.isInteger(b)) {
        throw new Error("Inputs must be integers.");
    }
    if (a < 0 || b < 0) {
        throw new Error("Inputs must be non-negative for standard GCD definition.");
    }

    // Base cases:
    // If b is 0, then a is the GCD
    while (b !== 0) {
        const remainder = a % b; // Calculate the remainder
        a = b;                   // Update a to be the previous b
        b = remainder;           // Update b to be the remainder
    }
    return a; // When b becomes 0, a holds the GCD
}

/**
 * Problem 2: Least Common Multiple (LCM) using GCD
 *
 * The LCM of two integers is the smallest positive integer that is divisible by
 * both numbers.
 *
 * Optimal Approach: Using GCD
 * There's a direct relationship between LCM and GCD:
 * LCM(a, b) = |a * b| / GCD(a, b)
 * This formula is efficient because GCD can be computed efficiently.
 *
 * @param {number} a The first non-negative integer.
 * @param {number} b The second non-negative integer.
 * @returns {number} The least common multiple of a and b.
 *
 * Time Complexity: O(log(min(a, b))) - Dominated by the GCD calculation.
 * Space Complexity: O(1) - Uses a constant amount of extra space.
 */
function lcm(a, b) {
    // Input validation similar to GCD
    if (typeof a !== 'number' || typeof b !== 'number' || !Number.isInteger(a) || !Number.isInteger(b)) {
        throw new Error("Inputs must be integers.");
    }
    if (a < 0 || b < 0) {
        throw new Error("Inputs must be non-negative for standard LCM definition.");
    }

    // Handle edge case where one or both numbers are 0.
    // The LCM of 0 and any number is 0.
    if (a === 0 || b === 0) {
        return 0;
    }

    // Calculate LCM using the formula: LCM(a, b) = (a * b) / GCD(a, b)
    // To prevent potential overflow issues with `a * b` for very large numbers,
    // it's sometimes safer to calculate `(a / gcd(a, b)) * b`.
    const commonDivisor = gcd(a, b);
    return (a / commonDivisor) * b;
}

/**
 * Problem 3: Sieve of Eratosthenes for finding prime numbers up to a limit
 *
 * The Sieve of Eratosthenes is an ancient algorithm for finding all prime numbers
 * up to a specified integer. It works by iteratively marking as composite (i.e., not prime)
 * the multiples of each prime, starting with the first prime number, 2.
 *
 * Optimal Approach: Sieve of Eratosthenes
 * 1. Create a boolean array `isPrime[0...limit]` and initialize all entries as true.
 *    `isPrime[0]` and `isPrime[1]` are marked false.
 * 2. Start from p = 2. While p*p <= limit:
 *    a. If `isPrime[p]` is true, then p is a prime.
 *    b. Mark all multiples of p (p*p, p*p + p, p*p + 2p, ...) up to `limit` as false.
 *       We start from p*p because smaller multiples (like 2*p, 3*p) would have already
 *       been marked by smaller primes (2, 3 respectively).
 *    c. Increment p.
 * 3. All numbers `i` for which `isPrime[i]` is true are prime numbers.
 *
 * @param {number} limit The upper bound (inclusive) up to which primes should be found.
 * @returns {number[]} An array containing all prime numbers up to `limit`.
 *
 * Time Complexity: O(limit * log(log(limit))) - This is a very efficient prime-finding algorithm.
 *                  The outer loop runs up to sqrt(limit). The inner loop marks multiples.
 *                  Each number is visited a number of times proportional to its number of prime factors.
 *                  The harmonic series sum property leads to this complexity.
 * Space Complexity: O(limit) - An array of size `limit + 1` is used to store primality information.
 */
function sieveOfEratosthenes(limit) {
    if (typeof limit !== 'number' || !Number.isInteger(limit) || limit < 0) {
        throw new Error("Limit must be a non-negative integer.");
    }

    // Create a boolean array 'isPrime[0...limit]' and initialize all entries as true.
    // A value in isPrime[i] will be true if i is prime, false otherwise.
    const isPrime = new Array(limit + 1).fill(true);

    // 0 and 1 are not prime numbers
    if (limit >= 0) isPrime[0] = false;
    if (limit >= 1) isPrime[1] = false;

    // Start checking from p = 2. We only need to check up to sqrt(limit)
    // because if a number n has a prime factor greater than sqrt(n), it must
    // also have a prime factor smaller than sqrt(n).
    for (let p = 2; p * p <= limit; p++) {
        // If isPrime[p] is still true, then it is a prime
        if (isPrime[p]) {
            // Mark all multiples of p as not prime.
            // Start from p*p because smaller multiples (2p, 3p, etc.) would
            // have already been marked by smaller primes.
            for (let i = p * p; i <= limit; i += p) {
                isPrime[i] = false;
            }
        }
    }

    // Collect all prime numbers
    const primes = [];
    for (let i = 2; i <= limit; i++) {
        if (isPrime[i]) {
            primes.push(i);
        }
    }

    return primes;
}

/**
 * Problem 4: Power Function (x^n) using Binary Exponentiation (Exponentiation by Squaring)
 *
 * Calculate x raised to the power of n (x^n).
 *
 * Optimal Approach: Binary Exponentiation (Exponentiation by Squaring)
 * This method drastically reduces the number of multiplications required, especially for large exponents.
 * It's based on the idea that:
 * - If n is even, x^n = (x^2)^(n/2)
 * - If n is odd, x^n = x * x^(n-1) = x * (x^2)^((n-1)/2)
 * This can be implemented iteratively by checking bits of n in binary representation.
 *
 * @param {number} base The base number (x).
 * @param {number} exp The exponent (n). Can be positive, negative, or zero.
 * @returns {number} The result of base^exp.
 *
 * Time Complexity: O(log(abs(exp))) - The number of multiplications is proportional to the number of bits in `exp`.
 * Space Complexity: O(1) - Uses a constant amount of extra space.
 */
function power(base, exp) {
    if (typeof base !== 'number' || typeof exp !== 'number' || !Number.isInteger(exp)) {
        throw new Error("Base must be a number and exponent must be an integer.");
    }

    // Handle special cases:
    if (exp === 0) {
        // x^0 = 1 for any x (including 0^0 which is typically 1 in combinatorics, though undefined in pure math)
        return 1;
    }
    if (base === 0) {
        // 0^n = 0 for n > 0
        return 0;
    }

    let result = 1;
    let currentBase = base;
    let currentExp = Math.abs(exp); // Work with positive exponent first

    // Iterate while currentExp is greater than 0
    while (currentExp > 0) {
        // If the current bit of exp is 1 (i.e., exp is odd)
        if (currentExp % 2 === 1) {
            result *= currentBase; // Multiply result by currentBase
        }
        currentBase *= currentBase; // Square the currentBase (e.g., x -> x^2 -> x^4 -> x^8)
        currentExp = Math.floor(currentExp / 2); // Halve the exponent (e.g., n -> n/2 -> n/4)
    }

    // If the original exponent was negative, the result is 1 / (base^|exp|)
    if (exp < 0) {
        return 1 / result;
    }

    return result;
}

/**
 * Problem 5: Nth Fibonacci Number
 *
 * The Fibonacci sequence is a series of numbers where each number is the sum of
 * the two preceding ones, usually starting with 0 and 1.
 * F(0) = 0, F(1) = 1, F(n) = F(n-1) + F(n-2) for n > 1.
 *
 * Optimal Approach: Iterative Dynamic Programming
 * This approach builds the sequence from the bottom up, storing only the last two
 * numbers needed to calculate the next, thus avoiding redundant calculations and
 * stack overflow issues of naive recursion.
 *
 * @param {number} n The index of the Fibonacci number to find (non-negative integer).
 * @returns {number} The Nth Fibonacci number.
 *
 * Time Complexity: O(n) - The loop runs n-1 times.
 * Space Complexity: O(1) - Uses a constant amount of extra space (two variables).
 *
 * Alternative (More optimal for very large N): Matrix Exponentiation (O(log n))
 * This approach is typically used for extremely large N (e.g., 10^18) where O(N) is too slow.
 * It involves raising a 2x2 matrix to the power of N using binary exponentiation for matrices.
 * This implementation focuses on the more common iterative DP.
 */
function nthFibonacci(n) {
    if (typeof n !== 'number' || !Number.isInteger(n) || n < 0) {
        throw new Error("Input 'n' must be a non-negative integer.");
    }

    if (n === 0) return 0; // F(0) = 0
    if (n === 1) return 1; // F(1) = 1

    let a = 0; // Represents F(i-2)
    let b = 1; // Represents F(i-1)
    let temp;

    // Start from i = 2 up to n
    for (let i = 2; i <= n; i++) {
        temp = a + b; // Calculate F(i) = F(i-2) + F(i-1)
        a = b;        // Update F(i-2) to F(i-1)
        b = temp;     // Update F(i-1) to F(i)
    }

    return b; // b will hold F(n)
}

/**
 * Problem 6: Combinations (N choose K)
 *
 * Calculates the number of ways to choose k items from a set of n items without regard to the order.
 * Denoted as C(n, k) or (n k).
 * Formula: C(n, k) = n! / (k! * (n-k)!)
 *
 * Optimal Approach: Pascal's Triangle (Row/Diagonal optimization) or Formula with pre-cancellation
 * Direct factorial calculation can lead to very large intermediate numbers, potentially exceeding
 * JavaScript's safe integer limit, even if the final result fits.
 * A more robust approach is to simplify the formula:
 * C(n, k) = (n * (n-1) * ... * (n-k+1)) / (k * (k-1) * ... * 1)
 * This involves k multiplications and k divisions, mitigating intermediate overflow.
 * Also, C(n, k) = C(n, n-k), so we can choose min(k, n-k) to reduce iterations.
 *
 * @param {number} n The total number of items (non-negative integer).
 * @param {number} k The number of items to choose (non-negative integer).
 * @returns {number} The number of combinations C(n, k).
 *
 * Time Complexity: O(k) - The loop runs k times (or min(k, n-k) times).
 * Space Complexity: O(1) - Uses a constant amount of extra space.
 */
function combinations(n, k) {
    if (typeof n !== 'number' || typeof k !== 'number' || !Number.isInteger(n) || !Number.isInteger(k)) {
        throw new Error("Inputs 'n' and 'k' must be integers.");
    }
    if (n < 0 || k < 0) {
        throw new Error("Inputs 'n' and 'k' must be non-negative.");
    }
    if (k > n) {
        return 0; // Cannot choose more items than available
    }
    if (k === 0 || k === n) {
        return 1; // C(n, 0) = 1, C(n, n) = 1
    }

    // Optimization: C(n, k) = C(n, n-k)
    // Choose the smaller value between k and n-k to minimize calculations
    k = Math.min(k, n - k);

    let result = 1;
    for (let i = 0; i < k; i++) {
        result = result * (n - i) / (i + 1);
    }

    // Ensure the result is an integer (due to floating point division, although it should be exact)
    return Math.round(result);
}


module.exports = {
    gcd,
    lcm,
    sieveOfEratosthenes,
    power,
    nthFibonacci,
    combinations,
};
```